src/TypeInterface.jl

```diff
@@ -0,0 +1,49 @@
+module TypeInterfaceModule
+
+# Must implement:
+
+# Is a value of your type valid?
+is_valid(x::T) where {T} = true #=::Bool=#
+is_valid_array(x::AbstractArray{T}) where {T} = all(is_valid, x) #=::Bool=#
+
+# Obtain the underlying number type
+function get_number_type(::Type{T}) where {T} #=::DataType=#
+    error("Base number type of your custom type is not defined")
+    return T
+end
+
+# Given an array of numbers, append all number constants stored in your type at a given index
+function append_number_constants!(
+    nvals::AbstractVector{BT}, idx::Int64, value::T
+) where {BT<:Number,T}
+    return idx
+end #=::Int64=#
+
+# Given an array of numbers, get all the number constants starting at a given index and store them in your type.
+function #=::Tuple{Int64,T}=#
+pop_number_constants(nvals::AbstractVector{BT}, idx::Int64, value::T) where {T,BT<:Number}
+    return (idx, value)
+end #=::Tuple{Int64,T}=#
```

Comment:
```suggestion
"""
    pop_number_constants(nvals, idx, value)

The inverse of `append_number_constants!`. Given an array `nvals` containing data,
fill `value` with `nvals` starting from `idx`.

Returns a tuple of the next index to read from, and the filled-in value.

!!! note
    In the case of a scalar `value`, this will simply return `(idx, value)` without affecting `nvals`.
"""
function pop_number_constants(nvals::AbstractVector{BT}, idx::Int64, value::T) where {T,BT<:Number}
    return (idx, value)
end
```